libcontainer/nsenter/nsexec.c

```diff
@@ -2,351 +2,350 @@
+	/* Send the child pid back to our parent */
+	len = snprintf(buf, JSON_MAX, "{\"pid\": %d}\n", childpid);
+	if (len < 0 || write(pipenum, buf, len) != len) {
 		kill(childpid, SIGKILL);
-		exit(1);
+		bail("unable to send a child pid");
 	}
 
 	exit(0);
 }
 
 static struct nsenter_config process_nl_attributes(int pipenum, char *data, int data_size)
 {
-	struct nsenter_config	config	    = {0};
-	struct nlattr		*nlattr;
-	int			payload_len;
-	int			start       = 0;
+	char *current = data;
+	struct nsenter_config config = { 0 };
 
 	config.consolefd = -1;
-	while (start < data_size) {
-		nlattr = (struct nlattr *)(data + start);
-		start += NLA_HDRLEN;
-		payload_len = nlattr->nla_len - NLA_HDRLEN;
-
-		if (nlattr->nla_type == CLONE_FLAGS_ATTR) {
-			config.cloneflags = readint32(data + start);
-		} else if (nlattr->nla_type == CONSOLE_PATH_ATTR) {
-			// get the console path before setns because it may
-			// change mnt namespace
-			config.consolefd = open(data + start, O_RDWR);
-			if (config.consolefd < 0) {
-				pr_perror("Failed to open console %s",
-					  data + start);
-				exit(1);
-			}
-		} else if (nlattr->nla_type == NS_PATHS_ATTR) {
-			// if custom namespaces are required, open all
-			// descriptors and perform setns on them
-			int	i, j;
-			int	nslen = num_namespaces(data + start);
-			int	fds[nslen];
-			char	*nslist[nslen];
-			char	*ns;
-			char	*saveptr = NULL;
-
-			for (i = 0; i < nslen; i++) {
-				char *str = NULL;
-
-				if (i == 0) {
-					str = data + start;
-				}
-				ns = strtok_r(str, ",", &saveptr);
-				if (ns == NULL) {
-					break;
-				}
-				fds[i] = open(ns, O_RDONLY);
-				if (fds[i] == -1) {
-					for (j = 0; j < i; j++) {
-						close(fds[j]);
-					}
-					pr_perror("Failed to open %s", ns);
-					exit(1);
+	while (current < data + data_size) {
+		struct nlattr *nlattr = (struct nlattr *)current;
+		size_t payload_len = nlattr->nla_len - NLA_HDRLEN;
+
+		/* Advance to payload. */
+		current += NLA_HDRLEN;
+
+		/* Handle payload. */
+		switch (nlattr->nla_type) {
+		case CLONE_FLAGS_ATTR:
+			config.cloneflags = readint32(current);
+			break;
+		case CONSOLE_PATH_ATTR:
+			/*
+			 * The context in which this is done (before or after we
+			 * join the other namespaces) will affect how the path
+			 * resolution of the console works. This order is not
+			 * decided here, but rather in container_linux.go. We just
+			 * follow the order given by the netlink message.
+			 */
+			config.consolefd = open(current, O_RDWR);
+			if (config.consolefd < 0)
+				bail("failed to open console %s", current);
+			break;
+		case NS_PATHS_ATTR:{
+				/*
+				 * Open each namespace path and setns it in the
+				 * order provided to us. We currently don't have
+				 * any context for what kind of namespace we're
+				 * joining, so just blindly do it.
+				 */
+				char *saveptr = NULL;
+				char *ns = strtok_r(current, ",", &saveptr);
+				int *fds = NULL, num = 0, i;
+				char **paths = NULL;
```

Comment:
I prefer them here because they're local to this case. Also, I'm planning on rewriting most of this because we need to do all of this handling in the first child (due to the fact that the PID namespace applies on children and the first child has to send the PID of the second).